### PR TITLE
ENG-2732: Add component contract relation

### PR DESCRIPTION
**Pull request by SmartMonkey-git**

### CODE REVIEW

File: tycho-indexer/migrations/2024-01-10_tycho_patch/up.sql [hunk 1]
```diff
@@ -61,6 +61,21 @@ DROP TRIGGER invalidate_previous_contract_code ON contract_code;
...
+    -- Timestamp this entry was inserted into this table.
+    "inserted_ts" timestamptz NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    -- Timestamp this entry was inserted into this table.
+    "modified_ts" timestamptz NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    PRIMARY KEY ("protocol_component_id", "contract_code_id")
+);
+
+CREATE INDEX IF NOT EXISTS idx_protocol_component_token_protocol_component_id ON protocol_component_holds_contract(protocol_component_id);
+
+CREATE INDEX IF NOT EXISTS idx_protocol_component_token_token_id ON protocol_component_holds_contract(contract_code_id);
```
Comment (tamaralipows) [inline tycho-indexer/migrations/2024-01-10_tycho_patch/up.sql]: `token_token_id`?

File: tycho-indexer/src/storage/postgres/protocol.rs [hunk 2]
```diff
@@ -248,6 +248,7 @@ where
             })
             .collect();
 
+        // establish component-ontract junction
```
Comment (tamaralipows) [inline tycho-indexer/src/storage/postgres/protocol.rs]:
```suggestion
        // establish component-contract junction
```

File: tycho-indexer/src/storage/postgres/protocol.rs [hunk 3]
```diff
@@ -298,6 +299,61 @@ where
             .execute(conn)
             .await?;
 
+        // establish component-ontract junction
```
Comment (tamaralipows) [inline tycho-indexer/src/storage/postgres/protocol.rs]:
```suggestion
        // establish component-contract junction
```

File: tycho-indexer/src/storage/postgres/protocol.rs [hunk 4]
```diff
@@ -298,6 +299,61 @@ where
...
+        let contract_addresses: HashSet<Address> = new
+            .iter()
+            .flat_map(|pc| pc.get_byte_contract_addresses())
+            .collect();
+
+        let pc_entity_contract_map = new
```
Comment (tamaralipows) [inline tycho-indexer/src/storage/postgres/protocol.rs]: protocol component entity? Why not just pc?

File: tycho-indexer/src/storage/postgres/protocol.rs [hunk 5]
```diff
@@ -298,6 +299,61 @@ where
...
+            .iter()
+            .flat_map(|pc| {
+                let pc_id = protocol_db_id_map
+                    .get(&(pc.id.clone(), pc.protocol_system.clone(), pc.chain))
+                    .expect("Could not find Protocol Component. Even though it should have."); //Because we just inserted the protocol systems, there should not be any missing.
```
Comment (tamaralipows) [inline tycho-indexer/src/storage/postgres/protocol.rs]: I think just "Could not find Protocol Component" should be enough. The expect is enough to imply it should have :)

File: tycho-indexer/src/storage/postgres/protocol.rs [hunk 6]
```diff
@@ -893,12 +949,22 @@ mod test {
...
-        let weth_id =
+        let (account_id_weth, weth_id) =
             db_fixtures::insert_token(conn, chain_id, WETH.trim_start_matches("0x"), "WETH", 18)
                 .await;
-        let usdc_id =
+        let (account_id_usdc, usdc_id) =
             db_fixtures::insert_token(conn, chain_id, USDC.trim_start_matches("0x"), "USDC", 6)
                 .await;
+
+        println!("account_id_weth: {}", account_id_weth);
```
Comment (tamaralipows) [inline tycho-indexer/src/storage/postgres/protocol.rs]: Unnecessary?

File: tycho-indexer/migrations/2024-01-10_tycho_patch/up.sql [hunk 7]
```diff
@@ -61,6 +61,21 @@ DROP TRIGGER invalidate_previous_contract_code ON contract_code;
...
+    -- Timestamp this entry was inserted into this table.
+    "inserted_ts" timestamptz NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    -- Timestamp this entry was inserted into this table.
+    "modified_ts" timestamptz NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    PRIMARY KEY ("protocol_component_id", "contract_code_id")
+);
+
+CREATE INDEX IF NOT EXISTS idx_protocol_component_contract_protocol_component_id ON protocol_component_holds_contract(protocol_component_id);
+
+CREATE INDEX IF NOT EXISTS idx_protocol_component_contract_contract_id ON protocol_component_holds_contract(contract_code_id);
```
Comment (tamaralipows) [inline tycho-indexer/migrations/2024-01-10_tycho_patch/up.sql]: Still not sure why we repeat the word contract here? Is it necessary?

File: tycho-indexer/migrations/2024-01-10_tycho_patch/up.sql [hunk 8]
```diff
@@ -61,6 +61,21 @@ DROP TRIGGER invalidate_previous_contract_code ON contract_code;
 
 DROP FUNCTION invalidate_previous_entry_contract_code();
 
+CREATE TABLE IF NOT EXISTS protocol_component_holds_contract(
+    "protocol_component_id" bigint REFERENCES protocol_component(id) ON DELETE CASCADE NOT NULL,
+    -- we don't allow a token to be deleted unless the protocol component was removed
```
Comment (tamaralipows) [inline tycho-indexer/migrations/2024-01-10_tycho_patch/up.sql]: Contract, not token?

File: tycho-indexer/src/storage/postgres/orm.rs [hunk 9]
```diff
@@ -478,6 +478,13 @@ impl ProtocolComponent {
     }
 }
 
+#[derive(Insertable)]
+#[diesel(table_name = protocol_component_holds_contract)]
+pub struct NewProtocolComponentHoldsContract {
```
Comment (dianacarvalho1) [inline tycho-indexer/src/storage/postgres/orm.rs:484]: The other one is called `NewProtocolHoldsToken` not `NewProtocolComponentHoldsToken` :/ I prefer ProtocolComponent tbh, but yeah, renamings right?

Comment (SmartMonkey-git) [reply tycho-indexer/src/storage/postgres/orm.rs:484]: Sorry, I'm not sure what your preferred action would be here.
1. Rename `NewProtocolComponentHoldsContract` to `NewProtocolHoldsContract` for consistency
2. Rename `NewProtocolHoldsToken` to `NewProtocolComponentHoldsToken` for better naming

I would prefer 2., if it doesn't cost a lot of time. WDYT?

Comment (dianacarvalho1) [reply tycho-indexer/src/storage/postgres/orm.rs:484]: yes I would prefer 2 too

**Pull request merged**